### PR TITLE
Handle as

### PR DESCRIPTION
**Pull request by CodeBaboon**
This provides a convenience for consumers to not have to always chain a Body method if they want the response (we'll do it for them). The `component.fetch` response is now a wrapper with a real window.Response inside, along with (optionally) the body content (if handle-as is set).

### CODE REVIEW

**Comment by CodeBaboon** [comment]
@dlockhart I'm curious on your thoughts as to whether you think there is more value to the consumer in adding something like this at the expense of not behaving in the same exact fashion as `window.fetch` would. I sort of like not necessarily having to check `ok` and then chain the appropriate `Body.whatever()` call as a consumer, but there is also a certain elegance in just returning a `window.Response` (though it is still there, just wrapped now).

File: d2l-fetch.html [hunk 1]
```diff
@@ -90,20 +96,47 @@
...
 					.then(function(token) {
 						this.authToken = token;
-						return window.fetch(this._getRequest());
+						return this._doFetch(this.handleAs);
 					}.bind(this));
 			},
 
+			_doFetch: function(handleAs) {
+				var self = this;
+				return window.fetch(self._getRequest())
```
Comment (dlockhart) [inline d2l-fetch.html]: What's going to happen in IE11... I don't think it natively supports `window.fetch`?

**Comment by dlockhart** [comment]
Yeah I definitely like sticking as closely to the underlying API as possible if we can.

So I guess the goal here is an alternative to `d2l-ajax` that uses the fetch API? Just thinking out loud... another idea would be to turn this into a behaviour that you can attach to your web components that simply wraps `window.fetch`. Then the API could be identical:

```
this.fetch( new Request(...) )
    .then(function(response) {
    });
```

**Comment by dlockhart** [comment]
Another idea would be to start with a simple promise-based function that turns a `Request` into a `Request` that has the right D2L auth headers added. Then consumers can actually use the native `window.fetch` directly.

e.g.
```
var authedRequest = d2lAuthifyRequest( new Request( ... ) )
    .then( function( request ) {
        return window.fetch( request );
    }.then( function( response ) {
    } );
```

File: d2l-fetch.html (hunk 1 above)
Comment (CodeBaboon) [inline d2l-fetch.html]: We are making use of a fetch polyfill for that scenario: https://github.com/BrightspaceUI/fetch/blob/master/bower.json#L14

**Comment by CodeBaboon** [comment]
The behavior idea is interesting though I'm not sure it makes sense to mixin this functionality to another component? I'm not sure. The second idea of a promise-based function maybe loses some of the caching/local storage goodness? Not sure.

But yeah the basic idea here is to replace `d2l-ajax` with something similar but that uses `window.fetch`. I think most pertinently I was sort of wondering if the idea of `fetch()` returning:

```
{
    response: Response, // this is a real window.fetch Response object
    body: body // this is optionally the .json(), .text(), etc value read from the Response object
}
```

Was better/worse than just returning the result of `window.fetch()`.

The downside to just returning `window.fetch()` is that to read the response the consumer needs to chain another `then` and call the right Body function, like:

```
.fetch()
 .then(function(response) {
    if (response.ok) {
        response.json()
            .then(function(body) {
                // do something with the body
            });
    } else {
        // handle what to do in the case of 404, 500, etc
    }
})
.catch(function(err) {
    // do something with the rejection
}
```

Basically this PR just wraps that intermediary checking of the `response.ok` and then calling the body method (when the user supplies a `handle-as` attrib value).

**Comment by dlockhart** [comment]
The main reason I like a behaviour (and don't really like `d2l-ajax`...) is that with the web component you end up in a weird situation when you want to make multiple calls, since the attributes retain state. I guess that can be both good and bad.

I just like that if you need to make multiple requests in parallel you can just go:
```
Promise.all([
    this.fetch( new Request( url1 ) ),
    this.fetch( new Request( url2 ) )
]).then( function( res1, res2 ) {
});
```

Or in sequence:
```
this.fetch( new Request( url1 ) )
   .then(function( res1 ) {
      return this.fetch( new Request( url2 ) );
   }).then(function( res2 ) {
   });
```

With the web component, it's a little trickier I think... you'd need a couple of them or with one you'd need to overwrite its attributes before making the second call.

**Comment by dlockhart** [comment]
But to answer your _actual_ question... I don't feel strongly either way. My preference I think would be to just return the native object, since that makes the documentation "just read Mozilla's docs".

**Comment by ryantmer** [comment]
Definitely :+1: to the native object - the extra `response.ok` is a bit meh, but ultimately I think the thinner this is, the better it is. `d2l-ajax` did what it needed to, but turned out to not make much sense (hindsight is 20/20 and all). I found one of the main downsides was that it attempted to re-implement `iron-ajax` fully, which was a lot of nitpicky stuff. Instead, I think it would have made more sense to just shim in the cookie auth/caching stuff with as little change as possible - whether by behavior, WC, or `d2lAuthifyRequest`-type-thing.

**Comment by CodeBaboon** [comment]
Yeah I should have updated this PR earlier. @dlockhart and I had some conversation to the side and I'm going to do a native component. In fact, I began work on it today. I'll put it up into a new repo as WIP so you guys can have a look before I get too far down the path again. And once we are cool that it is the right approach I'll close this PR (and probably delete this repo).
